docs/PowerAuth-Server-1.8.0.md

```diff
@@ -31,3 +31,38 @@ values supported by the WebAuthn protocol, serialized as a JSON array.
 A new column `status_reason` has been added to the `pa_operation` table.
 It provides optional details why the status changed.
 The value should be sent in the form of a computer-readable code, not a free-form text.
+
+### Idempotency-Key of Callback URL Events
+
+Callback URL Events now include an `Idempotency-Key` in the HTTP request header. It is a unique key to recognize retries
+of the same request.
+
+### New Database Table for Callback Events Monitoring
+
+A new `pa_application_callback_event` table has been created to monitor Callback URL Events. This change introduces
+the additional benefit of setting a retry strategy for individual Callback URL Events and monitoring the state of each
+dispatched event. The table contains following columns:
+- `id` - Event identifier, used also as the `Idempotency-Key`.
+- `application_callback_id` - Reference for corresponding Callback URL record in the `pa_application_callback` table.
+- `callback_data` - Data payload of the Callback URL Event.
+- `status` - Current state of the Callback URL Event.
+- `timestamp_created` - Creation timestamp of the Callback URL Event.
+- `timestamp_last_call` - Timestamp of the last time the Callback URL Event was sent.
+- `timestamp_next_call` - Timestamp of the next scheduled time to send the Callback URL Event.
+- `timestamp_delete_after` - Timestamp after which the Callback URL Event record should be deleted from the table.
+- `attempts` - Number of dispatch attempts made for the Callback URL Event.
+
+The `pa_application_callback_event` table comes with following indices:
+- `pa_application_callback_event_status_timestamp_next_call_idx` on `(status, timestamp_next_call)`,
```

Comment:
Doesn't time range filtering benefits from an index?

Reply:
Like, for this type of table and expected data  - I think the benefit of indexing are wipe out be need of additional index structure.

How many PENDING callback do you expect? with the granularity of the nextcall this will not drastically reduce the cost of the search. On the other hand DB will need to maintain index, the DB will need to select the proper algorithm,... When deleting records the index is also recalculated,...
Plus you are hitting larger part of index (because you are selecting older then something) - yes it could be faster, but again the maintenance....

So, If we expect hundreds of PENDING events (Or many very old PENDING records forcing to read far part of the table), yes, it will start paying off.  But from my POV it will not.



btw as far as I recall index should be constructed from the most distinctive column to the least - BTREE is still favourite selection algorithm not licenced by Oracle.

Reply:
good point with the timestamp granularity, the index probably won't have a significant effect in this case then